sgv2-docsapi/src/main/java/io/stargate/sgv2/docsapi/api/service/json/JsonConverter.java

```diff
@@ -0,0 +1,364 @@
+   */
+  public Map<String, Integer> generateColumnIndexMap(List<QueryOuterClass.ColumnSpec> columnSpecs) {
+    Map<String, Integer> colIndexes = new HashMap<>();
+    for (int index = 0; index < columnSpecs.size(); index++) {
+      QueryOuterClass.ColumnSpec col = columnSpecs.get(index);
+      colIndexes.put(col.getName(), index);
+    }
+    return colIndexes;
+  }
+
+  /**
+   * Takes a List of rows from a documents table, iterating over it and constructing a JSON object
+   * (or array) that is represented by those rows. Note: Because the system avoids
+   * read-before-write, it is possible for the same "path" in a document to contain two different
+   * sets of data from two different versions of the document. When this is encountered during the
+   * conversion, the version that has the latest writetime is accepted, and the older version is
+   * collected in the DeadLeafCollector to be purged in the background.
+   *
+   * @param rows Rows from gRPC
+   * @param collector a DeadLeafCollector implementation, to handle deletion of old and conflicting
+   *     data
+   * @param writeAllPathsAsObjects Instead of writing arrays such as [1, 2], write an object such as
+   *     {"0": 1, "1": 2}
+   * @param numericBooleans If these rows do not support boolean values and are using tinyint
+   *     instead
+   * @param maxDepth The rows' max depth (i.e. the highest pN value)
+   * @return the JsonNode represented by converting the @param rows
+   */
+  private JsonNode convertToJsonDoc(
+      List<QueryOuterClass.Row> rows,
+      List<QueryOuterClass.ColumnSpec> columns,
+      DeadLeafCollector collector,
+      boolean writeAllPathsAsObjects,
+      boolean numericBooleans,
+      int maxDepth) {
+    JsonNode doc = mapper.createObjectNode();
+    Map<String, Long> pathWriteTimes = new HashMap<>();
+    if (rows.isEmpty()) {
+      return doc;
+    }
+
+    Map<String, Integer> columnNameToIndex = generateColumnIndexMap(columns);
+
+    for (QueryOuterClass.Row row : rows) {
+      Long rowWriteTime = row.getValues(columnNameToIndex.get("writetime(leaf)")).getInt();
+      String rowLeaf = row.getValues(columnNameToIndex.get("leaf")).getString();
+      if (rowLeaf.equals(Constants.ROOT_DOC_MARKER)) {
+        continue;
+      }
+
+      String leaf = null;
+      JsonNode parentRef = null;
+      JsonNode ref = doc;
+
+      String parentPath = "$";
+
+      for (int i = 0; i < maxDepth; i++) {
+        String p = row.getValues(columnNameToIndex.get("p" + i)).getString();
+        String nextP =
+            i < maxDepth - 1 ? row.getValues(columnNameToIndex.get("p" + (i + 1))).getString() : "";
+        boolean endOfPath = nextP.equals("");
+        boolean isArray = p.startsWith("[");
+        boolean nextIsArray = nextP.startsWith("[");
+
+        if (isArray) {
+          // This removes leading zeros if applicable
+          p = "[" + Integer.parseInt(p.substring(1, p.length() - 1)) + "]";
```

Comment:
this could be part of the docs api utils?

Reply:
I haven't ported that class yet...is there already a place where it's been partially ported or not yet?